modules/compress/src/main/java/org/apache/ignite/internal/processors/compress/CompressionProcessorImpl.java

```
@@ -0,0 +1,345 @@
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.CompactablePageIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.PageIO;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+import static org.apache.ignite.configuration.PageCompression.SKIP_GARBAGE;
+import static org.apache.ignite.internal.util.GridUnsafe.NATIVE_BYTE_ORDER;
+
+/**
+ * Compression processor.
+ */
+public class CompressionProcessorImpl extends CompressionProcessor {
+    /** */
+    static boolean testMode = false;
+
+    /** A bit more than max page size. */
+    private static final int THREAD_LOCAL_BUF_SIZE = 18 * 1024;
+
+    /** */
+    private final ThreadLocalByteBuffer tmp = new ThreadLocalByteBuffer(THREAD_LOCAL_BUF_SIZE);
+
+    /**
+     * @param ctx Kernal context.
+     */
+    public CompressionProcessorImpl(GridKernalContext ctx) {
+        super(ctx);
+    }
+
+    /**
+     * @param cap Capacity.
+     * @return Direct byte buffer.
+     */
+    static ByteBuffer allocateDirectBuffer(int cap) {
+        return ByteBuffer.allocateDirect(cap).order(NATIVE_BYTE_ORDER);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void checkPageCompressionSupported() throws IgniteCheckedException {
+        if (testMode)
+            return;
+
+        if (!U.isLinux())
+            throw new IgniteCheckedException("Currently page compression is supported only for Linux.");
+
+        FileSystemUtils.checkSupported();
+    }
+
+    /** {@inheritDoc} */
+    @Override public ByteBuffer compressPage(
+        ByteBuffer page,
+        int pageSize,
+        int blockSize,
+        PageCompression compression,
+        int compressLevel
+    ) throws IgniteCheckedException {
+        assert compression != null;
+        assert U.isPow2(pageSize): pageSize;
+        assert page.position() == 0 && page.limit() == pageSize;
+
+        if (!U.isPow2(blockSize))
```

Can we move these checks to the processor startup? It would be better to fail early on activation rather than at checkpoint.

---

Agree. Done.